### ISSUE TITLE
Notification issue: Mailgun mail address format in filter cause rejected mails (Former Name: Postfix - Header differences cause mail rejection of OMV notifications)

### ISSUE THREAD

**Issue by that-coding-human** [openmediavault/openmediavault#832]
I tried to setup my mail notification using mailgun. I am using mailgun as it is a convenient mail relay for automatically sending server notifications through smtp as well as API, which makes it very flexible and a perfect match if you want ONE single point of contact for sending notifications from different parts of your infrastructure. (Just as an explanation why I am not using gmail :-) )

I setup the Notifications with the following settings:
SMTP Server: smtp.mailgun.org
SMTP Port: 465
Encryption Mode: SSL/TLS
Sender E-Mail: postmaster@sandbox1234567890.mailgun.org
Authentication required: true
Username: postmaster@sandbox1234567890.mailgun.org
Password: TottallyObviouslyNotMyPassword

Recipient:
myAdress@someDomain.com

For small setups and testing (before going productive) the free offering of mailgun is sufficient, but it requires to register and verify the recipient adresses. I did this, as the recipient for this setup is my own adress.

When pressing "send a test email", the following gets placed in the postfix qeue.
I used the command: `postcat -q 3E5B542110D `
#############################################################################

```
*** ENVELOPE RECORDS deferred/3/3E5B542110D ***
message_size:             470             353               2               0             470               0
message_arrival_time: Tue Oct 20 10:24:30 2020
create_time: Tue Oct 20 10:24:30 2020
named_attribute: rewrite_context=local
sender_fullname: root
sender: postmaster@sandbox1234567890.mailgun.org
named_attribute: notify_flags=1
original_recipient: openmediavault-notification@localhost.localdomain
done_recipient: openmediavault-notification@localhost.localdomain
*** MESSAGE CONTENTS deferred/3/3E5B542110D ***
Received: by myNAS.local (Postfix, from userid 0)
        id 3E5B542110D; Tue, 20 Oct 2020 10:24:30 +0200 (CEST)
To: myAdress@someDomain.com
Subject: Test message from myNAS.local
Date: Tue, 20 Oct 2020 10:24:30 +0200
From: postmaster@sandbox1234567890.mailgun.org
Message-Id: <20201020082430.3E5B542110D@myNAS.local>

This is an automatically generated mail message from the system notification service running on myNAS.local.
*** HEADER EXTRACTED deferred/3/3E5B542110D ***
named_attribute: dsn_orig_rcpt=rfc822;myAdress@someDomain.com
original_recipient: myAdress@someDomain.com
recipient: myAdress@someDomain.com
*** MESSAGE FILE END deferred/3/3E5B542110D ***
```

#############################################################################

The mail is not being delivered, but there is a message in the mail.log

```
Oct 20 19:52:32 myNAS postfix/qmgr[14956]: 3E5B542110D: from=<postmaster@sandbox1234567890.mailgun.org>, size=470, nrcpt=2 (queue active)
Oct 20 19:52:34 myNAS postfix/smtp[17442]: 3E5B542110D: host smtp.mailgun.org[3.93.221.84] said: 421 Domain sandbox1234567890.mailgun.org is not allowed to send: Sandbox subdomains are for test purposes only. Please add your own domain or add the address to authorized recipients in Account Settings. (in reply to DATA command)
Oct 20 19:52:36 myNAS postfix/smtp[17442]: 3E5B542110D: to=<myAdress@someDomain.com>, relay=smtp.mailgun.org[34.224.137.116]:465, delay=34086, delays=34082/0.07/3/0.46, dsn=4.0.0, status=deferred (host smtp.mailgun.org[34.224.137.116] said: 421 Domain sandbox1234567890.mailgun.org is not allowed to send: Sandbox subdomains are for test purposes only. Please add your own domain or add the address to authorized recipients in Account Settings. (in reply to DATA command))
```

At first this looks like a clear message, but the adress myAdress@someDomain.com IS a verified account, and I CAN send mails to it with mailgun. Sadly the mailgun logs do not contain this kind of rejected smtp requests, so I went back to my OMV and tried to exclude things step by step. At first I verified that there is no issue in the recipients replacement. The recipient_canonical looks good:
```
root myAdress@someDomain.com
admin myAdress@someDomain.com
@myNAS.local myAdress@someDomain.com
```
The main.cf looks good as well:
```
# This file is auto-generated by openmediavault (https://www.openmediavault.org)
# WARNING: Do not edit this file, your changes will get lost.

compatibility_level = 2
mydomain = local
mydestination =
inet_interfaces = localhost
relayhost = smtp.mailgun.org:465
sender_canonical_maps = regexp:/etc/postfix/sender_canonical
sender_bcc_maps = hash:/etc/postfix/sender_bcc
recipient_canonical_maps = hash:/etc/postfix/recipient_canonical
recipient_bcc_maps = hash:/etc/postfix/recipient_bcc
smtp_sasl_auth_enable = yes
smtp_sasl_security_options = noanonymous
smtp_sasl_password_maps = hash:/etc/postfix/sasl_passwd
smtp_sasl_mechanism_filter =
smtp_tls_wrappermode = yes
smtp_tls_security_level = encrypt
smtp_tls_CAfile = /etc/ssl/certs/ca-certificates.crt
smtp_tls_CApath = /etc/ssl/certs
smtp_header_checks = regexp:/etc/postfix/smtp_header_checks
alias_maps = hash:/etc/aliases
transport_maps = hash:/etc/postfix/transport
```

After I got more familiar with how OMV handles its mail notifications, I tried to send a mail just using CLI just with postfix, for narrowing down my issue and eliminating OMV from the equation.

I used the following command for sending a mail:
`echo "This is a test" | mail -s "Testing" myAdress@someDomain.com`
This worked like a charm. Mail got sent, received an I got a success message like this in the log file:
`Oct 20 09:58:37 myNAS postfix/smtp[14959]: 9F83142111B: to=<myAdress@someDomain.com>, relay=smtp.mailgun.org[54.184.250.231]:465, delay=2.2, delays=0.02/0/1.6/0.58, dsn=2.0.0, status=sent (250 Great success)
`
This puzzled me, as I was at first thinking I had an issue with mailgun or with my configuration, but now it looked like the root cause is in OMV. I repeated this test with a non-existing recipient, in order to provoke a rejection at mailgun, and compare the queue contents.
As expected the mail got rejected, with a similar message as above (no validated recipient). So I had a look at the queue, and got this:
(using the command: `postcat -q A38CC42111B` 
```
*** ENVELOPE RECORDS deferred/A/A38CC42111B ***
message_size:             451             353               2               0             451               0
message_arrival_time: Tue Oct 20 10:38:50 2020
create_time: Tue Oct 20 10:38:50 2020
named_attribute: rewrite_context=local
sender_fullname: root
sender: postmaster@sandbox1234567890.mailgun.org
named_attribute: notify_flags=1
original_recipient: openmediavault-notification@localhost.localdomain
done_recipient: openmediavault-notification@localhost.localdomain
*** MESSAGE CONTENTS deferred/A/A38CC42111B ***
Received: by myNAS.local (Postfix, from userid 0)
        id A38CC42111B; Tue, 20 Oct 2020 10:38:50 +0200 (CEST)
To: invalid-adress@someDomain.com
Subject: Testing
**MIME-Version: 1.0
Content-Type: text/plain; charset="UTF-8"
Content-Transfer-Encoding: 8bit**
Message-Id: <20201020083850.A38CC42111B@myNAS.local>
Date: Tue, 20 Oct 2020 10:38:50 +0200 (CEST)
**From: root <postmaster@sandbox1234567890.mailgun.org>**

This is test 5
*** HEADER EXTRACTED deferred/A/A38CC42111B ***
named_attribute: encoding=8bit
named_attribute: dsn_orig_rcpt=rfc822;invalid-adress@someDomain.com
original_recipient: invalid-adress@someDomain.com
recipient: invalid-adress@someDomain.com
*** MESSAGE FILE END deferred/A/A38CC42111B ***
```

Besides the different placement of some attributes in the header (which I consider uncritical), I found some differences (marked bold). The header contained additional information for MIME-Version, Content-Type and Content-Transfer.
In addition the From: field was formatted different. 
Actually it looks like some mail providers (like mailgun) reject mails if they are missing certain fields. The fact that mailgun then sends out a confusing error message is a different topic.

I tried to figure out why the header fields are so different and found:  openmediavault/deb/openmediavault/usr/share/openmediavault/notification/postfix-runner.sh 

In this script OMV rewrites the header, of notification mails. I assume this is the root cause of this behaviour.

I have the feeling this is as far as I can get at this point of researching. 

What do you think about this? Did you have any special thoughts and considerations why the OMV mail header differs so much from the regular postfix header? Is my research on this correct, or did I go down the wrong rabbithole?

I suggest to enhance the OMV mail header in order to match the regular postfix behaviour again (e.g. by adding additional parameters and sed in the postfix-runner.sh), or by not completely "rewriting" the mail but modifying it.

I am very curious about this, and just started to scratch the postfix surface. Even though I was only able to test with mailgun, it does not seem to be unlikely that there are more mail providers out there that block the OMV notifications.

Maybe there are others out here who can confirm this behaviour?

### Versions report
OMV Version: 5.5.12-1
Kernel: Linux 5.8.0-0-bpo.2-amd64

**Comment by votdev** [comment]
The script `/usr/share/openmediavault/notification/postfix-runner.sh` is called by postfix while processing mails to be able to support other notification types beside email. So it does not modify the emails to be sent.

I'm no postfix expert, so i need to have a look into this issue.

**Comment by votdev** [comment]
OMV does no magic when sending emails, it also uses the `mail` command and the [PHP equivalent function](https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/php/openmediavault/email.inc#L66). This PHP function requires additional header information that are added to the envelope to successful submit an email. I think the `mail` command is doing the same.

**Comment by that-coding-human** [comment]
Thank for this incredibly fast answer! As OMV also uses `mail`, I am wondering why the headers look different. I could not put my hands on the exact root cause yet. Usually postfix only handles the mails some other program gives to it, so the header might be set outside of postfix by `mail`, which is called in a different way when being called by CLI or the PHP function, as they might not be identical. 
I am currently investigating this from two sides:
1.) What is the cause for the different headers? (as I suspect this to influence the handling of the mail provider)
2.) Where and why exactly does mailgun block mails based on their header information (besides the recipient). 

I do not expect any help here regarding 2.), but if you have any further findings regarding 1.), I am eager to learn about this. I am not a postfix / mail expert either .... yet. :-) Maybe we can solve this puzzle together. If my observed behaviour can be fixed by a workaround or configuration adjustment, I can offer to write an how-to article, or provide input for the OMV documentation as contribution. I really searched a lot and could not find any article that could explain or help me with this. 
(sadly I do not know PHP enough to contribute anything regarding the Web UI, but I will continue looking into postifx and the shell part).

**Comment by votdev** [comment]
Is it possible to you to test the https://github.com/openmediavault/openmediavault/pull/834/files pull-request? Simply modify the file `/usr/share/php/openmediavault/email.inc`.

**Comment by that-coding-human** [comment]
I tested it by uncommenting the original line and replacing it with the changed line from the pull request.
```
        public function send() {
                // Prepare additional email headers.
                $headers = [];
                // Set date in RFC 2882 format.
                $headers[] = sprintf('Date: %s', date('r'));
                // Append the sender address.
                if (!empty($this->from))
                        $headers[] = sprintf('From: %s', $this->from);
                return mail($this->to, $this->subject, $this->message,
                        // implode("\r\n", $headers));
                        implode(PHP_EOL, $headers));

```
The behaviour did not change. The mails in the queue look the same, and the server keeps rejecting them, while sending with `mail `on the command line still works.

It is a good Idea it might have something to do with the line breaks, but it looks like this was not the root cause. My bet still is on either the header fields for content-type etc., or this small difference in the From header field.

MIME-Version: 1.0
Content-Type: text/plain; charset="UTF-8"
Content-Transfer-Encoding: 8bit
From: root <postmaster@sandbox1234567890.mailgun.org>

Are you aware of how to manipulate the header of mails that are in the postfix queue? If we have a way to just add these fields to the header of this message (that is stuck in the queue), and resend it, we can easily test this. I did not find a good way yet.

**Comment by that-coding-human** [comment]
I found and solved my issue, and it is so incredibly dumb that I need to take deep breath, and remind myself that we are in the year 2020. I will post a comprehensive answer, just wanted to let you know that the root cause is not OMV related. There might be something we can do here to prevent this again. Hang on. Just wanted to let you know, so you do not waste any more time on investigating and going down this rabbithole.

**Comment by that-coding-human** [comment]
TL:DR at the end.

First of all I need to apologize on behalf of mailgun and whoever wrote their WebUI portal and filter for verified addresses.
I mentioned at the beginning that mailgun offers a free plan, which can be used to send mails to authorized addresses using their API and SMTP (even offering different smtp users at your choice for your different applications.). But you are only allowed to send to verified addresses. 

To verify an address you need to enter it in their webUI and they send a verification mail with a code to this address that you have to enter in their UI again, in order to verify the address. No rocket science. Keep this in mind for later.

Fast forward to today.
I found a way to manipulate the queue and extract the full mail header and body to a file and resend the exact same mail, after manipulating the header in the file.

For further reference a link for postfix queue manipulation overview:
https://linuxhint.com/postfix_mail_queue_management/

And a link to a q&a about how to extract and manipulate postfix queued mails:
https://serverfault.com/questions/623211/postfix-change-sender-in-queued-messages

I went with option 3., in order to halt the queue, extract the mail to a file, manipulate it, and then using sendmail to send it and remove the original queue.
```
postsuper -h queueid
postcat -qbh queueid > tempfile.eml
vi tempfile.eml
sendmail $sender $recipient < tempfile.eml
postsuper -d queueid

```
I tried to take one of the failed mails and add the mentioned header information, and resend it, but it failed again. 
So I went through every field, and either manipulated or removed it completely, but it just kept failing, even after I took the original mail, for which I forced the error by using an invalid recipient. As this also failed I started to rethink my whole assumption.

I went through all my tests and my old logs from the past days, in order to find any small difference that might provide any hint on the root cause. In fact all sent e-mails are deleted from the queue, so I did not have the exact content of these headers anymore. So I wanted to recreate everything from scratch, and try to get hold of the header of successfully sent e-mails. I had a gut feeling that maybe the header of the successful mails is different from the header of the mail that I forced to fail (with an invalid address), in order to get hold of the content.

To verify the original success, I retried sending the message. After that I wanted to send it again, but halt the queue programmatically or disconnect the network in order to keep the mail in the queue.
`echo "myMessage" | mail -s "Test" myAdress@someDomain.com
`
THIS FAILED WITH THE EXACT SAME MESSAGE LIKE ALL THE OTHERS !!

After taking a deep breath I went through my command history in order to find the EXACT command that worked, and I found a little difference.

The command that worked was:
`echo "This is a test" | mail -s "Testing" myadress@somedomain.com`

As I replaced the mail address in this github issue with a placeholder, as I did not want to shout around my private addresses in public, the replacement in my original issue post did not reflect that detail.
In a rush I obviously wrote the recipient address in ALL LOWERCASE!
I quickly checked the verification mails that I got from mailgun and found that they sent the mails to my address with an all lowercase recipient as well. Usually mail addresses should not be handled case sensitive, right? I always use camel case for my technical mail addresses to make it more readable for me, and it should not matter technically.

I retried with:
`echo "This is a test" | mail -s "Testing" myadress@somedomain.com`
and it worked! 
I went to the OMV webUI, changed my Mail address to all lowercase, and immediately after the notifications kept pouring in my mailbox!
I deleted the verified address at mailgun and re-verified it, just to make absolutely sure.
Obviously mailgun takes the given addresses in their WebUI, and **converts** them to all lowercase, no matter how you entered it. Then it is verified and given to a filter ... that acts CASE SENSITIVE !!
As I always use camel case, and copy paste this around, this hit me. The non matching header fields from OMV and `mail` led me down the wrong path in the rabbit hole, while my successful test was just coincidence because I manually entered the recipient in lowercase.

I would never have expected a company in 2020 to handle mail addresses with a case sensitive filter and in addition transform user input to all lowercase before. Wow.

This is absolutely no root cause at OMV, even if it appeared like one to me at first. I deeply apologize if you feel like this was a waste of time. 

As others might be using mailgun as well, there is only one  enhancement that could be done to OMV in order to technically  prevent this. You could also convert all recipients / sender to all lowercase before sending them. I know it is ugly, and it feels like cleaning up the mess of others, but maybe it saves some time to people who are using providers with similar implementations.  A possible disadvantage of this would be to hit other issues for providers who use case sensitive exact match filters for the strings of mail addresses without transforming them to lowercase before ;-) 
This is not an enhancement request, just a suggestion to think about. I am not sure either if this really helps.

As most of the examples rely on gmail, I suggest to add a Warning box in the OMV documentation regarding this issue with mailgun and maybe other providers, so people are aware to try lowercase / uppercase mail addresses as part of their trouble shooting, before opening an issue here:
https://github.com/openmediavault/openmediavault-docs/blob/master/administration/general/notifications.rst

I will write a small article of how I configured notification for me, and I will mark this lesson learned big and bold. 

I think I will look both sides now before crossing a one-way street. This is how much trust I have left now in the application of common rules. :-)
Hopefully mailbox fixes this on their side as well. I dropped them a message, but I have no big hopes for it.


TL:DR - Mailgun transforms the validated addresses to all lowercase, before adding them to their filter, which acts case sensitive. Using a camel case address as recipient will trigger the filter to reject it. This is the root cause to my issue. I tried with my regular camel case address, and the one time it worked I just entered the address in lowercase manually in a hurry.

**Comment by votdev** [comment]
Great to hear that it is not the fault of OMV. IMO you should not think this was waste of time, you learned much exactly because of that.
